hybris/common/q/linker_main.cpp

```diff
@@ -0,0 +1,844 @@
+    if (args.argc <= 1 || !strcmp(args.argv[1], "--help")) {
+      async_safe_format_fd(STDOUT_FILENO,
+         "Usage: %s program [arguments...]\n"
+         "       %s path.zip!/program [arguments...]\n"
+         "\n"
+         "A helper program for linking dynamic executables. Typically, the kernel loads\n"
+         "this program because it's the PT_INTERP of a dynamic executable.\n"
+         "\n"
+         "This program can also be run directly to load and run a dynamic executable. The\n"
+         "executable can be inside a zip file if it's stored uncompressed and at a\n"
+         "page-aligned offset.\n",
+         args.argv[0], args.argv[0]);
+      exit(0);
+    }
+    exe_to_load = args.argv[1];
+    __libc_shared_globals()->initial_linker_arg_count = 1;
+  }
+
+  // store argc/argv/envp to use them for calling constructors
+  g_argc = args.argc - __libc_shared_globals()->initial_linker_arg_count;
+  g_argv = args.argv + __libc_shared_globals()->initial_linker_arg_count;
+  g_envp = args.envp;
+  __libc_shared_globals()->init_progname = g_argv[0];
+
+  // Initialize static variables. Note that in order to
+  // get correct libdl_info we need to call constructors
+  // before get_libdl_info().
+  sonext = solist = solinker = get_libdl_info(kLinkerPath, tmp_linker_so);
+  g_default_namespace->add_soinfo(solinker);
+  init_link_map_head(*solinker, kLinkerPath);
+
+  ElfW(Addr) start_address = linker_main(args, exe_to_load);
+
+  INFO("[ Jumping to _start (%p)... ]", reinterpret_cast<void*>(start_address));
+
+  // Return the address that the calling assembly stub should jump to.
+  return start_address;
+}
+#endif
+
+
+static void generate_tmpsoinfo(soinfo& tmp_linker_so) {
+  Dl_info dlinfo;
+  int ret = dladdr((void *)generate_tmpsoinfo, &dlinfo);
+  if(!ret) {
+    PRINT("can't get self info:ret\n", ret);
+    exit(-1);
+  }
+  ElfW(Addr) linker_addr = (ElfW(Addr))dlinfo.dli_fbase;
+  DEBUG("get self base adder=%p\n", linker_addr);
+
+  ElfW(Ehdr)* elf_hdr = reinterpret_cast<ElfW(Ehdr)*>(linker_addr);
+  ElfW(Phdr)* phdr = reinterpret_cast<ElfW(Phdr)*>(linker_addr + elf_hdr->e_phoff);
+
+  tmp_linker_so.base = linker_addr;
+  tmp_linker_so.size = phdr_table_get_load_size(phdr, elf_hdr->e_phnum);
+  tmp_linker_so.load_bias = get_elf_exec_load_bias(elf_hdr);
+  tmp_linker_so.dynamic = nullptr;
+  tmp_linker_so.phdr = phdr;
+  tmp_linker_so.phnum = elf_hdr->e_phnum;
+  tmp_linker_so.set_linker_flag();
+
+  DEBUG("tmp_linker_so's load_bias=%p \n", tmp_linker_so.load_bias);
+
+  // Prelink the linker so we can access linker globals.
+  if (!tmp_linker_so.prelink_image()) {
+    PRINT("can't prelink self:ret\n");
+  };
+
+  //tmp_linker_so.call_constructors();
+}
+
+#if (ANDROID_VERSION_MAJOR >= 10)
+bionic_tls* __allocate_temp_bionic_tls() {
+  size_t allocation_size = __BIONIC_ALIGN(sizeof(bionic_tls), PAGE_SIZE);
+  void* allocation = mmap(nullptr, allocation_size,
```

Comment:
> > any updates about this PR? it seems not merged to the repo?
> 
> I am not aware of anybody having successfully brought up UI with this, which i would see as a requirement.
> Also there is still the question about the TLS variables.
> 
> EDIT: sorry, apparently the question was only submitted right now since i forgot to press "submit review"... (see below)

i just want to have a try, make PostMarket OS with libhybris -Android 10.0  supported.